include/dmlc/optional.h

```diff
@@ -107,6 +108,8 @@ class optional {
   }
   /*! \brief whether this object is holding a value */
   explicit operator bool() const { return !is_none; }
+  /*! \brief alternate, more-explicit check of object value */
+  const bool IsNone() const { return is_none; }
```

Comment:
After some research, I see that the existing code includes an explicit conversion from an optional&lt;T&gt; to bool.  Since C++11, these can be considered in many common "contextual conversions to bool," including the controlling expression of if-statements.  Note the absence of 'operator' in the IsNone() method I'm proposing, so I don't see any issues in disturbing the existing conversion behavior.

My motivation in adding IsNone() is one of coding clarity.  In the case of optional&lt;int&gt; , the coder is often protected from leaving off a .value() call (which returns an int, so different from bool).  In the case of optional&lt;bool&gt;, no such protection exists.  I was hoping to steer the usage to the more explicit (and in my mind more understandable):

if ( my_optional_bool.IsNone() )
    my_optional_bool = some_good_default_value;

Without the IsNone() method, this must appear as:

if ( !my_optional_bool )
    my_optional_bool = some_good_default_value;

Someone not intimately familiar with the optional&lt;T&gt; class might think "if my_optional_bool is false, then set it to some_good_default_value."

Bottom line, I can remove this from the PR if it's getting in the way of its acceptance.  It would be my preference to keep it, but it represents no fundamental new functionality.